spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Container.scala

```diff
@@ -40,19 +42,70 @@ import scala.reflect.ClassTag
 abstract class Container[A <: Activity : ClassTag,
     B <: Activity : ClassTag, T: ClassTag](
   implicit ev: TensorNumeric[T]) extends AbstractModule[A, B, T] {
-
   // list of sub modules
-  val modules: ArrayBuffer[AbstractModule[Activity, Activity, T]]
-  = ArrayBuffer[AbstractModule[Activity, Activity, T]]()
+  val modules: ArrayBuffer[IModule[Activity, Activity, T]]
+  = ArrayBuffer[IModule[Activity, Activity, T]]()
+
+  // Return empty list as compile do nothing by default.
+  def compilingPath(): List[Node[IModule[Activity, Activity, T]]] = {
+    List()
+  }
+
+  def gatherFinalResult(values: List[Activity]): Activity = {
+    if (values.isEmpty) {
+      return null
+    }
+    if (values.length == 1) {
+      return values(0)
+    } else {
+      val t = new Table()
+      values.foreach {v =>
+        t.insert(v)
+      }
+      return t
+    }
+  }
+
+  protected def doCompile(executionNodes: List[ModuleNode[T]]): Unit = {
+    var i = 0
+    while (i < executionNodes.length) {
+      val node = executionNodes(i)
+      val preNodes = node.prevNodes
+      val inputShapes = if (preNodes.isEmpty) {
+        if (node.element.getInputShape() == null) {
+            throw new RuntimeException("The first layer should explicitly declare inputShape")
+        } else {
+          List(node.element.getInputShape())
+        }
+      } else {
+        preNodes.map{_.element.getOutputShape()}.toList
+      }
+      node.element.build(gatherFinalResult(inputShapes))
+      i += 1
+    }
+  }
+
+  final def compile(): Unit = {
```

Comment:
I don't think this should be a container API. Say we add model.compile in the optimizer, then all user need to define model by first declare a container

Reply:
`compile` would be transparent to the user and should be invoked internally.